File: search/interpolation_search/Java/InterpolationSearch.java
import java.util.Arrays;

public class InterpolationSearch {
    
    private static final int NOT_FOUND = -1;
    
    static int interpolationSearch(int[] arr, int x) {
        
        // Keep track of the two indexes 
        int lo = 0;
        int hi = (arr.length - 1);
        
        while (lo <= hi && x >= arr[lo] && x <= arr[hi]) {
            
            // Calculate the position to check
            int pos = lo + (((hi-lo) /
                    (arr[hi]-arr[lo]))*(x - arr[lo]));

            // The key was found! 
            if (arr[pos] == x){
                return pos; 
            }

            // If x is larger, x is in upper part 
            if (arr[pos] < x) {
                lo = pos + 1;
            }

            // If x is smaller, x is in the lower part 
            else {
                hi = pos - 1;
            }
        }
        
        // The key was not found!
        return NOT_FOUND;
    }

    // Driver method  
    public static void main(String[] args)
    {
        int key = 18;
        int[] test = {10, 12, 13, 16, 18, 19, 20, 21, 22, 23, 24, 33, 35, 42, 47};
        int index = interpolationSearch(test, key);
        System.out.println("The Sample Array: " + Arrays.toString(test));
        System.out.println("The Test Key: " + key);
        System.out.println("The Index Of The Key: " + index);
    }
}

File: search/interpolation_search/cpp/interpolation_search.cpp
#include <bits/stdc++.h>
using namespace std;
bool interpolation_search(int arr[],int size,int element){
	int low=0;
	int high=size-1;
	int pos=0;
	while(low <= high and element >= arr[low] and element <= arr[high]){
		//pos = l + [ (x-arr[l])*(h-l) / (arr[h]-arr[L]) ]
		pos=low+((element-arr[low])*(high-low))/(arr[high]-arr[low]);
		if(arr[pos] == element ){
			return true;
		} 
		if(arr[pos]>element)high=pos-1;
		else low=pos+1;
	}
	return false;
}
int main(){
	int element=10;
	int arr[]={1,5,6,13,29,30,39,42};
	int n=sizeof(arr)/sizeof(arr[0]);
	bool found=interpolation_search(arr,n,element);
	if(found){
		cout<<element<<" is present in array :)\n";
	}
	else cout<<element<<" is not present in array :(\n";
	element=29;
	found=interpolation_search(arr,n,element);
	if(found){
		cout<<element<<" is present in array :)\n";
	}
	else cout<<element<<" is not present in array :(\n";
}